sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -84,86 +94,97 @@ static unsigned _count_char(const char *s, char c)
 {
-    uint8_t code;
-    switch (err) {
-    case -EACCES:
-        code = COAP_CODE_FORBIDDEN;
-        break;
-    case -ENOENT:
-        code = COAP_CODE_PATH_NOT_FOUND;
-        break;
-    default:
-        code = COAP_CODE_INTERNAL_SERVER_ERROR;
-    };
-    DEBUG("gcoap_fileserver: Rejecting error %d (%s) as %d.%02d\n", err, strerror(err),
-          code >> 5, code & 0x1f);
+    uint8_t code = err;
+    if (err < 0) {
+        switch (err) {
+            case -EACCES:
+                code = COAP_CODE_FORBIDDEN;
+                break;
+            case -ENOENT:
+                code = COAP_CODE_PATH_NOT_FOUND;
+                break;
+            default:
+                code = COAP_CODE_INTERNAL_SERVER_ERROR;
+        };
+        DEBUG("gcoap_fileserver: Rejecting error %d (%s) as %d.%02d\n", err, strerror(err),
+            code >> 5, code & 0x1f);
+    }
     return gcoap_response(pdu, buf, len, code);
 }
 
-static void _calc_szx2(coap_pkt_t *pdu, size_t reserve, struct requestdata *request)
+static void _calc_szx2(coap_pkt_t *pdu, size_t reserve, coap_block1_t *block2)
 {
     assert(pdu->payload_len > reserve);
     size_t remaining_length = pdu->payload_len - reserve;
     /* > 0: To not wrap around; if that still won't fit that's later caught in
      * an assertion */
-    while ((coap_szx2size(request->szx2) > remaining_length) && (request->szx2 > 0)) {
-        request->szx2--;
-        request->blocknum2 <<= 1;
+    while ((coap_szx2size(block2->szx) > remaining_length) && (block2->szx > 0)) {
+        block2->szx--;
+        block2->blknum <<= 1;
     }
 }
 
-static ssize_t gcoap_fileserver_file_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len,
-                                             struct requestdata *request)
+static ssize_t _get_file(coap_pkt_t *pdu, uint8_t *buf, size_t len,
+                         struct requestdata *request,
+                         gcoap_fileserver_entry_t *resource)
 {
+    (void)resource;
+    int err;
     uint32_t etag;
-    int err = stat_etag(request->namebuf, &etag);
-    if (err < 0) {
-        return gcoap_fileserver_errno_handler(pdu, buf, len, err);
+    coap_block1_t block2 = {.offset = 0,
+                            .blknum = 0,
+                            .szx = CONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX,
+                            .more = false};
+    {
+        struct stat stat;
+        if ((err = vfs_stat(request->namebuf, &stat)) < 0) {
+            return gcoap_fileserver_error_handler(pdu, buf, len, err);
+        }
+        stat_etag(&stat, &etag);
     }
-
-    if (request->etag_sent && etag == request->etag) {
+    if (request->options.exists.block2 && !coap_get_block2(pdu, &block2)) {
+        return gcoap_fileserver_error_handler(pdu, buf, len, COAP_CODE_BAD_OPTION);
+    }
+    if (request->options.exists.if_match &&
+        memcmp(&etag, &request->options.if_match, request->options.if_match_len)) {
```

Comment:
This could need a comment explaining what the if_match does.

Reply:
I am going to add an explanation what `If-Match` and `If-None-Match` are supposed to be used for in this example, in `fileserver.h`.

The content of an `If-Match` option, sent by a client, is a previously received Etag. What it expresses is the following: The client wants to make sure that the resource it wants to update has not been updated by someone else, since the client had last received that Etag. The `If-Match` option can have 0 length which basically means: If the resource exists, I want my request to be processed. If the `If-Match` option does not match, the request failed to meet its precondition.

The `If-None-Match` option has length 0. It expresses that a client wants to create a resource only if it has not been created already.